File: sunday_catechism/sunday_catechism/report/student_birthdays_by_week/student_birthdays_by_week.js
frappe.query_reports["Student Birthdays by Week"] = {
	filters: [
		{
			fieldname: "year",
			label: __("Year"),
			fieldtype: "Int",
			default: new Date().getFullYear(),
			reqd: 1,
		},
		{
			fieldname: "class",
			label: __("Class"),
			fieldtype: "Link",
			options: "Class",
		},
		{
			fieldname: "active_only",
			label: __("Active Students Only"),
			fieldtype: "Check",
			default: 1,
		},
	],

	onload(report) {
		const METHOD =
			"sunday_catechism.sunday_catechism.print.birthday_list.generate_birthday_list";

		// Server-rendered PDF grouped by week (Frappe reports can't carry a custom Jinja
		// print, so we stream our own — same pattern as register printing).
		report.page.add_inner_button(__("Print (Grouped by Week)"), () => {
			const common = {
				year: frappe.query_report.get_filter_value("year"),
				class_name: frappe.query_report.get_filter_value("class") || "",
				active_only: frappe.query_report.get_filter_value("active_only") ? 1 : 0,
			};
			const base = `/api/method/${METHOD}`;
			// preview=1 -> inline (Content-Disposition: inline) so it renders in the iframe;
			// the download link omits it so the browser saves the file.
			const preview_url = `${base}?${$.param({ ...common, preview: 1 })}`;
			const download_url = `${base}?${$.param(common)}`;

			const d = new frappe.ui.Dialog({
				title: __("Birthday List Preview"),
				size: "extra-large",
				fields: [
					{
						fieldtype: "HTML",
						fieldname: "preview",
						options: `<iframe src="${preview_url}" title="${__(
							"Birthday List Preview"
						)}" style="width:100%; height:75vh; border:1px solid var(--border-color, #d1d8dd); border-radius:var(--border-radius, 6px);"></iframe>`,
					},
				],
				primary_action_label: __("Download PDF"),
				primary_action() {
					window.open(download_url);
				},
			});

			d.show();
		});
	},
};

File: sunday_catechism/sunday_catechism/report/student_birthdays_by_week/student_birthdays_by_week.py
import calendar

import frappe
from frappe import _
from frappe.utils import getdate


def execute(filters=None):
	filters = filters or {}
	year = int(filters.get("year") or getdate().year)
	is_leap = calendar.isleap(year)

	student_filters = {"date_of_birth": ["is", "set"]}
	if filters.get("active_only"):
		student_filters["active"] = 1
	if filters.get("class"):
		student_filters["class"] = filters.get("class")

	students = frappe.get_all(
		"Student",
		filters=student_filters,
		fields=["admission_no", "full_name", "date_of_birth", "class"],
	)

	data = []
	for s in students:
		dob = getdate(s.date_of_birth)
		month, day = dob.month, dob.day
		# Feb 29 birthdays are celebrated on Mar 1 in non-leap years.
		if month == 2 and day == 29 and not is_leap:
			month, day = 3, 1
		birthday = getdate(f"{year}-{month:02d}-{day:02d}")
		data.append(
			{
				"birthday": birthday,
				"full_name": s.full_name,
				"student_class": s.get("class"),
			}
		)

	# Order chronologically by the birthday.
	data.sort(key=lambda r: (r["birthday"], r["full_name"] or ""))

	return get_columns(), data


def get_columns():
	return [
		{
			"fieldname": "birthday",
			"label": _("Birthday"),
			"fieldtype": "Date",
			"width": 110,
		},
		{
			"fieldname": "full_name",
			"label": _("Student"),
			"fieldtype": "Data",
			"width": 200,
		},
		{
			"fieldname": "student_class",
			"label": _("Class"),
			"fieldtype": "Link",
			"options": "Class",
			"width": 150,
		},
	]
